### PR TITLE
Add benchmark for UnnestOperator

### PR DESCRIPTION
**Pull request by phd3**
CC @sopel39

### CODE REVIEW

**Comment by martint** [comment]
@cla-bot check

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+        public List<Page> getPages()
+        {
+            return pages;
+        }
+
+        private static List<Page> createInputPages(int positionsPerPage, List<Type> types, InputGenerator inputGenerator, double primitiveNullsUnnest, double rowNullsUnnest)
+        {
+            ImmutableList.Builder<Page> pages = ImmutableList.builder();
+            int numPages = TOTAL_POSITIONS / positionsPerPage;
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: nit: `pageCount` avoid the abbreviation in the variable name and reads better.

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+            if (outputPage != null) {
+                outputPages.add(outputPage);
+            }
+        }
+
+        return outputPages.build();
+    }
+
+    /**
+     * Helper class for random input generation.
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Remove this comment. It doesn't really add any info that's not obvious from the code below.

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+     */
+    public static class InputGenerator
+    {
+        private static final int INT_UPPER_LIMIT = 1_000_000_000;
+
+        private Random random = new Random(456);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: This can be final

Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: What's the significance of `456`?

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        private final int minNestedCardinality;
+        private final int maxNestedCardinality;
+
+        // upper bound on length of strings generated for VARCHAR elements
+        private final int maxStringLength;
+
+        public InputGenerator(int minNestedCardinality, int maxNestedCardinality, int maxStringLength)
+        {
+            checkArgument(minNestedCardinality >= 0 && maxNestedCardinality >= 0 && minNestedCardinality <= maxNestedCardinality, "cardinality bounds are invalid");
+            checkArgument(maxStringLength >= 0, "invalid string length");
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Be more explicit about what's invalid. Say something like "max string length must be >= 0"

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+         * indicates the percentage for Row elements.
+         *
+         * The null percentages are not propagated inside a MapBlock's keyBlock. Everything
+         * is always non-null inside the returned keyBlock.
+         *
+         * The set S of valid input for {@code type} can be defined as:
+         * S = {VARCHAR, INTEGER, ROW(x), ARRAY(x), MAP(x, x)} where x belongs to S
+         *
+         */
+        public Block produceBlock(Type type, int entries, double primitiveElementNulls, double rowElementNulls)
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: May be rename `primitiveElementNulls` and `rowElementNulls` to `xxxRatio` to make it more obvious what they represent (without having to look at the docs)

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+                }
+                else {
+                    VARCHAR.writeString(builder, generateRandomString(maxStringLength));
+                }
+            }
+
+            return builder.build();
+        }
+
+        private Block produceIntBlock(int entries, double primitiveNulls)
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Rename to `primitiveNullsRatio` (see above)

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+                }
+                else {
+                    INTEGER.writeLong(builder, random.nextInt(INT_UPPER_LIMIT));
+                }
+            }
+
+            return builder.build();
+        }
+
+        private Block produceArrayBlock(ArrayType arrType, int entries, double primitiveNulls, double rowElementNulls)
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Don't abbreviate variable names (`arrType`). Also rename the other variables as per the earlier comments.

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        {
+            BlockBuilder builder = arrType.createBlockBuilder(null, 100);
+            Type elementType = arrType.getElementType();
+
+            for (int i = 0; i < entries; i++) {
+                int arrLength = minNestedCardinality + random.nextInt(maxNestedCardinality - minNestedCardinality);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Don't abbreviate variable names (`arrLength`).

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+                fieldBlocks[i] = produceBlock(rowType.getTypeParameters().get(i), nonNullCount, primitiveNulls, rowElementNulls);
+            }
+
+            return RowBlock.fromFieldBlocks(entries, Optional.of(rowIsNull), fieldBlocks);
+        }
+
+        public static String generateRandomString(int length)
+        {
+            byte[] array = new byte[length];
+            new Random().nextBytes(array);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Why not use the `Random` instance in this class?

Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: The random bytes are not necessarily going to be a valid UTF-8 string. It'd be more robust to either 1) generate a string out of simple letters and numbers or 2) generate random codepoints and assemble the string from that, instead.

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java (hunk 3 above)
Comment (phd3) [reply presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: added `final` keyword and removed parameter.

File: presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java (hunk 5 above)
Comment (phd3) [reply presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java]: Agreed, changed the variable names.

**Comment by martint** [comment]
Merged, thanks! Travis failures are unrelated.

**Pull request merged**